### PR TITLE
create LSP23 LinkedContractsFactory

### PR DESCRIPTION
**Pull request by skimaharvey**
# What does this PR introduce?

Draft of le `LSP23`

### CODE REVIEW

File: LSPs/LSP-23-LinkedContractsFactory.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+## Motivation
+
+The ability to deploy linked contracts and execute post-deployment modules simplifies the process of deploying complex contract systems. This standard aims to provide a unified interface for deploying linked contracts and executing post-deployment logic.
+
+In the context of multi-chain Ethereum smart contract development, there are situations where a contract, used by an another contract, needs to be deployed at the same addresses across multiple chains. This presents a complex challenge as it requires deterministic address generation for the contracts across different chains. The `LSP23` standard addresses this issue by generating the salt for contract deployment within the function, ensuring a fully decentralized deployment process. This approach not only guarantees consistent contract addresses across multiple chains but also prevents address squatting, enhancing the security and integrity of the deployment process.
```
Comment (JeneaVranceanu) [inline LSPs/LSP-23-LinkedContractsFactory.md:37]: `used by an another contract` -> `used by another contract`

Comment (JeneaVranceanu) [inline LSPs/LSP-23-LinkedContractsFactory.md:37]: Just wanted to clarify. What "function" are you talking about? 
The part of the sentence is "by generating the salt for contract deployment within the function".

Comment (skimaharvey) [reply LSPs/LSP-23-LinkedContractsFactory.md:37]: thanks for the feedback. totally reviewed this part to avoid repetition

**Review by b00ste — APPROVED** [review]

File: LSPs/LSP-23-LinkedContractsFactory.md [hunk 2]
```diff
@@ -0,0 +1,760 @@
...
+  - [Methods](#methods)
+    - [deployContracts](#deploycontracts)
+    - [deployERC1167Proxies](#deployerc1167proxies)
+- [Implementation](#implementation)
+- [Interface Cheat Sheet](#interface-cheat-sheet)
+- [Copyright](#copyright)
+
+## Simple Summary
+
+Deploying smart contracts that need to interact with each other can be complicated. This is especially true when these contracts need to know each other's details at the time they are created. The LSP23 standard simplifies this process by providing a unified way to deploy such interdependent contracts. It also allows for additional actions to be taken automatically after the contracts are deployed, making the whole process more streamlined and less error-prone.
```
Comment (CJ42) [inline LSPs/LSP-23-LinkedContractsFactory.md:27]: small suggestions

```suggestion
Deploying smart contracts that need to interact with each other can be complicated. This is especially true when these contracts need to know each other's details at the time they are created. The LSP23 standard simplifies this process by providing a unified way to deploy such interdependent contracts. It also allows for additional actions to be taken automatically after the contracts have been deployed, streamlining the whole process and making it less error-prone.
```

File: LSPs/LSP-23-LinkedContractsFactory.md [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+
+## Abstract
+
+The LSP23 Linked Contracts Factory standard introduces a unified interface for deploying interdependent smart contracts, also known as linked contracts. These contracts often require each other's addresses at the time of deployment, creating a circular dependency. LSP23 addresses this by allowing for the deployment of a primary contract and one or more secondary contracts, linking them together. The standard also supports the deployment of contracts as `ERC1167` minimal proxies and enables optional post-deployment modules for executing additional logic. This serves to simplify and standardize the deployment of complex contract systems.
```
Comment (CJ42) [inline LSPs/LSP-23-LinkedContractsFactory.md:31]:
```suggestion
The LSP23 Linked Contracts Factory standard introduces an unified interface for deploying interdependent smart contracts, also known as **linked contracts**. These contracts often require each other's addresses at the time of deployment, creating a circular dependency. LSP23 addresses this by allowing for the deployment of a primary contract and one or more secondary contracts, linking them together. The standard also supports the deployment of contracts as [ERC1167](https://www.google.com/search?q=ERC1167&oq=ERC1167&aqs=chrome..69i57.2294j0j1&sourceid=chrome&ie=UTF-8) minimal proxies and enables optional post-deployment modules for executing additional logic. This serves to simplify and standardize the deployment of complex contract systems.
```

File: LSPs/LSP-23-LinkedContractsFactory.md [hunk 4]
```diff
@@ -0,0 +1,760 @@
...
+
+## Motivation
+
+While there are various methods for deploying smart contracts in the current ecosystem, the challenge of deploying interdependent contracts in a streamlined and secure manner remains. Custom solutions often exist, but they can be cumbersome and prone to errors. LSP23 aims to standardize this process, reducing the complexity and potential for mistakes. By generating the salt for contract deployment within the function, the standard also ensures a fully decentralized deployment process. This not only guarantees consistent contract addresses across multiple chains but also serves as a preventive measure against address squatting, enhancing the overall security and integrity of the deployment process.
```
Comment (CJ42) [inline LSPs/LSP-23-LinkedContractsFactory.md:35]: Should we specify CREATE2 here too?

```suggestion
While there are various methods for deploying smart contracts in the current ecosystem, the challenge of deploying interdependent contracts in a streamlined and secure manner remains. Custom solutions often exist, but they can be cumbersome and error prone. LSP23 aims to standardize this process, reducing the complexity and potential for mistakes. By generating the CREATE2 salt for contract deployment within the function, the standard also ensures a fully decentralized deployment process. This not only guarantees consistent contract addresses across multiple chains but also serves as a preventive measure against _"address squatting"_, enhancing the overall security and integrity of the deployment process.
```

File: LSPs/LSP-23-LinkedContractsFactory.md [hunk 5]
```diff
@@ -0,0 +1,760 @@
...
+
+## Specification
+
+### Linked Contracts Factory Setup
+
+#### Deployment
+
+Before the deployment of the `LSP23LinkedContractsFactory` on any network, people should make sure that the Nick Factory is deployed on the same network.
```
Comment (CJ42) [inline LSPs/LSP-23-LinkedContractsFactory.md:43]:
```suggestion
Before the deployment of the `LSP23LinkedContractsFactory` on any network, people should make sure that the [Nick Factory](https://yamenmerhi.medium.com/nicks-method-ethereum-keyless-execution-168a6659479c) is deployed on the same network.
```

File: LSPs/LSP-23-LinkedContractsFactory.md [hunk 6]
```diff
@@ -0,0 +1,760 @@
...
+
+##### Nick Factory Deployment
+
+The Nick Factory should be located at this address `0x4e59b44847b379578588920ca78fbf26c0b4956c` on the network. If there is no code on this address, it means that the contract is not deployed yet.
```
Comment (CJ42) [inline LSPs/LSP-23-LinkedContractsFactory.md:47]:
```suggestion
The Nick Factory contract should be located at this address `0x4e59b44847b379578588920ca78fbf26c0b4956c` on the network. If there is no code on this address, it means that the contract is not deployed yet.
```

File: LSPs/LSP-23-LinkedContractsFactory.md [hunk 7]
```diff
@@ -0,0 +1,760 @@
...
+
+To deploy, the following raw transaction should be broadcasted to the network `0xf8a58085174876e800830186a08080b853604580600e600039806000f350fe7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe03601600081602082378035828234f58015156039578182fd5b8082525050506014600cf31ba02222222222222222222222222222222222222222222222222222222222222222a02222222222222222222222222222222222222222222222222222222222222222` after funding the deployer address: `0x3fab184622dc19b6109349b94811493bf2a45362` with `gasPrice (100 gwei) * gasLimit (100000)`.
+
+Check [Nick's Factory repository](https://github.com/Arachnid/deterministic-deployment-proxy/tree/master) for more information.
+
+##### LSP23LinkedContractsFactory Deployment
+
+After the deployment of Nick Factory on the network, the `LSP23LinkedContractsFactory` can be reproduced at the standardized address given sending the same salt and bytecode.
```
Comment (CJ42) [inline LSPs/LSP-23-LinkedContractsFactory.md:55]:
```suggestion
After the deployment of the Nick Factory contract on the network, the `LSP23LinkedContractsFactory` can be reproduced at the standardized address given sending the same salt and bytecode.
```

File: LSPs/LSP-23-LinkedContractsFactory.md [hunk 8]
```diff
@@ -0,0 +1,760 @@
...
+
+In order to create the UniversalFactory contract, one should send a transaction to the [Nick Factory] address with data field equal to [salt](#standardized-salt) + [bytecode](#standardized-bytecode).
```
Comment (CJ42) [inline LSPs/LSP-23-LinkedContractsFactory.md:57]: Is this a mistake? Or this intended?

```suggestion
In order to create the `LSP23LinkedContractsFactory` contract, one should send a transaction to the [Nick Factory] address with data field equal to [salt](#standardized-salt) + [bytecode](#standardized-bytecode).
```

File: LSPs/LSP-23-LinkedContractsFactory.md [hunk 9]
```diff
@@ -0,0 +1,760 @@
...
+
+The address produced should be equal to `0x2300000A84D25dF63081feAa37ba6b62C4c89a30`.
+
+#### UniversalFactory Configuration
```
Comment (CJ42) [inline LSPs/LSP-23-LinkedContractsFactory.md:61]: Same shouldn't the title be different to be `LSP23LinkedContractsFactory` ✅ ? Not `UniversalFactory` ❌ ?

**Review by CJ42 — APPROVED** [review]
LGTM! 🚀  Small suggestions added. Also recommend to add any specific "Requirements" under each method if there is any (like we do with the other LSP specs documents)

**Comment by frozeman** [comment]
Merged preliminary

consider it DRAFT

**Pull request merged**